[Skia] Implement canvas ellipse

carlosgcampos opened the pull request:
#### cef2f7b37e4a6819e0a83b24a053d0c4c6b4e855
<pre>
[Skia] Implement canvas ellipse
<a href="https://bugs.webkit.org/show_bug.cgi?id=269569">https://bugs.webkit.org/show_bug.cgi?id=269569</a>

Reviewed by Nikolas Zimmermann.

Move the implementation from PathSkia::add(PathArc) to a new helper
method that can be used for PathSkia::add(PathEllipse). Also remove the
angle normalization, because angles are already normalized by
CanvasPath.

* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::addEllipse):
(WebCore::PathSkia::add):
* Source/WebCore/platform/graphics/skia/PathSkia.h:

Canonical link: <a href="https://commits.webkit.org/274974@main">https://commits.webkit.org/274974@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2cfd84896200dc8a717285197068e6fecab363fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40552 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19565 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42930 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43104 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36641 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42859 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22524 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16895 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33645 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41126 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16507 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34959 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14229 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14305 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35932 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44378 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36770 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36261 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39999 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15367 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12598 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38325 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16986 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17037 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5383 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16630 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/PathSkia.cpp:
```diff
@@ -164,9 +150,20 @@ void PathSkia::add(PathClosedArc closedArc)
+    if (ellipse.rotation) {
+        AffineTransform transform;
+        transform.translate(ellipse.center.x(), ellipse.center.y()).rotateRadians(ellipse.rotation);
+        AffineTransform inverseTransform = transform.inverse().value();
```
s/AffineTransform/auto/

nikolaszimmermann replied:
value_or(AffineTransform::identity) -- there are degenerate cases. maybe you even need to bail out for non-invertible case? value() is not safe here.

carlosgcampos replied:
I can add an assert, but the transform is always invertible, it's checked by the caller in`CanvasPath::ellipse()`

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/PathSkia.cpp:
```diff
@@ -164,9 +150,20 @@ void PathSkia::add(PathClosedArc closedArc)
+void PathSkia::add(PathEllipse ellipse)
 {
-    notImplemented();
+    if (ellipse.rotation) {
```
I personally would prefer to move the m_elementStream->add(ellipse) to the top, then

```
if (!ellipse.rotation) {
    addEllipse(...)
    return;
}

auto transform = AffineTransform::makeTranslation(...);
```

to avoid indenting the long branch.

carlosgcampos replied:
I don't know what the element stream is, is it ok to call that before adding the platform ellipse?

nikolaszimmermann reviewed (approved):
r=me, with a few minor comments.

carlosgcampos committed:
[Skia] Implement canvas ellipse
https://bugs.webkit.org/show_bug.cgi?id=269569

Reviewed by Nikolas Zimmermann.

Move the implementation from PathSkia::add(PathArc) to a new helper
method that can be used for PathSkia::add(PathEllipse). Also remove the
angle normalization, because angles are already normalized by
CanvasPath.

* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::addEllipse):
(WebCore::PathSkia::add):
* Source/WebCore/platform/graphics/skia/PathSkia.h:

Canonical link: https://commits.webkit.org/274974@main
